### PR TITLE
Add support for syntax highlighting in PyREPL

### PR DESCRIPTION
**Pull request by ambv**

### CODE REVIEW

File: Lib/_pyrepl/utils.py [hunk 1]
```diff
@@ -41,25 +85,82 @@ def unbracket(s: str, including_content: bool = False) -> str:
...
     chars: CharBuffer = []
     char_widths: CharWidths = []
 
     if not buffer:
         return chars, char_widths
 
-    for c in buffer:
+    while colors and colors[0].span.end < start_index:
+        # move past irrelevant spans
+        colors.pop(0)
```
Comment (pablogsal) [inline Lib/_pyrepl/utils.py:152]: I know this list is not big enough that it matters but it would recommend modifying this to figure out the cutting index and then slicing once

File: Lib/_pyrepl/utils.py [hunk 2]
```diff
@@ -73,5 +174,19 @@ def disp_str(buffer: str) -> tuple[CharBuffer, CharWidths]:
...
+
+        if colors and colors[0].span.end == i:  # current color ends now
+            post_color = TAG_TO_ANSI["SYNC"]
+            colors.pop(0)
+
+        chars[-1] = pre_color + chars[-1] + post_color
+        pre_color = ""
+        post_color = ""
+
+    if colors and colors[0].span.start < i and colors[0].span.end > i:
```
Comment (pablogsal) [inline Lib/_pyrepl/utils.py:186]: This approach assumes the next call to disp_str() will properly handle the color continuation. If any buffer modification happens between calls, or if the next call uses different parameters, you'll get incorrect highlighting no?

Maybe this is a no problem but I would recommend to either add some defensive check or an assert

```
SYNC"]

# Then in the next call, verify consistency
if pending_color_span and colors and colors[0].span == pending_color_span.span:
    # Spans match, continue highlighting
    pre_color = TAG_TO_ANSI[colors[0].tag]
else:
    …
```

File: Lib/_pyrepl/utils.py [hunk 3]
```diff
@@ -73,5 +174,19 @@ def disp_str(buffer: str) -> tuple[CharBuffer, CharWidths]:
         else:
             chars.append(c)
             char_widths.append(str_width(c))
+
+        if colors and colors[0].span.end == i:  # current color ends now
+            post_color = TAG_TO_ANSI["SYNC"]
+            colors.pop(0)
+
+        chars[-1] = pre_color + chars[-1] + post_color
```
Comment (pablogsal) [inline Lib/_pyrepl/utils.py:182]: Not sure if this matters but this being in a tight loop may be a performance problem

File: Lib/test/test_pyrepl/test_reader.py [hunk 4]
```diff
@@ -355,3 +362,60 @@ def test_setpos_from_xy_for_non_printing_char(self):
         reader, _ = handle_all_events(events)
         reader.setpos_from_xy(8, 0)
         self.assertEqual(reader.pos, 7)
+
+    def test_syntax_highlighting_basic(self):
```
Comment (pablogsal) [inline Lib/test/test_pyrepl/test_reader.py:366]: I would recommend adding some extra test for edge cases and failure modes such as incomplete code blocks, syntax errors and other known failure conditions so we know how this will behave

**Comment by ambv** [comment]
Moved on to python/cpython#131562.
